packages/fincaps/src/offerSigner.js

```diff
@@ -0,0 +1,167 @@
+const Agoric = {
+  Bech32MainPrefix: 'agoric',
+};
+
+/**
+ * @template {Record<string, unknown>} T
+ * @typedef {{[P in keyof T]: Exclude<T[P], undefined>;}} AllDefined
+ */
+
+/**
+ * Concise way to check values are available from object literal shorthand.
+ * Throws error message to specify the missing values.
+ *
+ * from @agoric/internal
+ *
+ * @template {Record<string, unknown>} T
+ * @param {T} obj
+ * @throws if any value in the object entries is not defined
+ * @returns {asserts obj is AllDefined<T>}
+ */
+export const assertAllDefined = obj => {
+  const missing = [];
+  for (const [key, val] of Object.entries(obj)) {
+    if (val === undefined) {
+      missing.push(key);
+    }
+  }
+  if (missing.length > 0) {
+    Fail`missing ${q(missing)}`;
+  }
+};
+
+/** @param {string} mnemonic */
+export const walletFromMnemonic = async mnemonic => {
+  const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {
+    prefix: Agoric.Bech32MainPrefix,
+    hdPaths: [hdPath(Agoric.CoinType, 0)],
+  });
+  const [signer] = await wallet.getAccounts();
+  return { wallet, signer };
+};
+
+/**
+ * @typedef {{
+ *   registry?: import('@cosmjs/proto-signing').Registry;
+ *   accountNumber?: number;
+ *   sequence?: number;
+ *   chainId?: string;
+ *   fee?: import('@cosmjs/amino').StdFee;
+ *   memo?: string;
+ *   timeoutHeight?: bigint,
+ * }} SignerOpts
+ */
+
+/**
+ * @param {DirectSecp256k1HdWallet} wallet
+ * @param {import('@cosmjs/proto-signing').AccountData} signer
+ * @param {import('@cosmjs/proto-signing').EncodeObject[]} messages
+ * @param {SignerOpts} opts
+ */
+export const sign = async (wallet, signer, messages, opts) => {
+  const {
+    accountNumber = Fail`accountNumber missing`,
+    chainId = Fail`chainId missing`,
+    sequence = Fail`sequence missing`,
+    fee = Fail`fee missing`,
+    memo,
+    timeoutHeight,
+    registry = SwingsetRegistry,
+  } = opts;
+  const pubkey = encodePubkey(encodeSecp256k1Pubkey(signer.pubkey));
+  /** @type {import('@cosmjs/proto-signing').TxBodyEncodeObject} */
+  const txBodyEncodeObject = {
+    typeUrl: '/cosmos.tx.v1beta1.TxBody',
+    value: {
+      messages: messages,
+      memo,
+      // @ts-expect-error signingstargateclient.ts uses bigint
+      timeoutHeight,
+    },
+  };
+  const txBodyBytes = registry.encode(txBodyEncodeObject);
+  const authInfoBytes = makeAuthInfoBytes(
+    [{ pubkey, sequence }],
+    fee.amount,
+    Number(fee.gas),
+    undefined,
+    undefined,
+  );
+  const signDoc = makeSignDoc(
+    txBodyBytes,
+    authInfoBytes,
+    chainId,
+    accountNumber,
+  );
+  const { signature, signed } = await wallet.signDirect(
+    signer.address,
+    signDoc,
+  );
+  return TxRaw.fromPartial({
+    bodyBytes: signed.bodyBytes,
+    authInfoBytes: signed.authInfoBytes,
+    signatures: [fromBase64(signature.signature)],
+  });
+};
+
+/** @param {import('./secret-tool').PassKey} item */
+const fromPassKey = async item => {
```

Comment:
What is `PassKey`? It seems like `secret-tool.js` wasn't checked in.... Does this read from a YubiKey? Or does secret-tool just get a secret from `process.env`?

Reply:
it's there: [secret-tool.js](https://github.com/dckc/finquick/blob/ag-offer-sign/packages/fincaps/src/secret-tool.js)

It lets you access the [FreeDesktop Secret Sevice][1], analogous to [keychain](https://en.wikipedia.org/wiki/Keychain_(software)).

from the package readme:

> ## Plugins with POLA: Google Sheets, Desktop Secrets, sqlite3
>
> Each of these plugins provdes POLA-shaped access:
>
> - ...
> - `secret-tool.js` - `SecretTool.makePasskey(attrs1)`
>  makes a `PassKey`. From there, `E(p).subKey(attrs2)`
>  attenuates access to items with `{ ...attrs2, ...attrs1 }`.

[1]: https://specifications.freedesktop.org/secret-service/latest/